File: implementation/scipamato/scipamato-core-persistence-jooq/src/test/java/ch/difty/scipamato/core/persistence/search/JooqSearchOrderServiceTest.java
package ch.difty.scipamato.core.persistence.search;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import ch.difty.scipamato.common.persistence.paging.PaginationContext;
import ch.difty.scipamato.core.entity.search.SearchCondition;
import ch.difty.scipamato.core.entity.search.SearchOrder;
import ch.difty.scipamato.core.entity.search.SearchOrderFilter;
import ch.difty.scipamato.core.persistence.AbstractServiceTest;

@ExtendWith(MockitoExtension.class)
@SuppressWarnings({ "ResultOfMethodCallIgnored", "OptionalGetWithoutIsPresent" })
public class JooqSearchOrderServiceTest extends AbstractServiceTest<Long, SearchOrder, SearchOrderRepository> {

    private static final String LC = "de";

    private JooqSearchOrderService service;

    @Mock
    private SearchOrderRepository repoMock;
    @Mock
    private SearchOrderFilter     filterMock;
    @Mock
    private PaginationContext     paginationContextMock;
    @Mock
    private SearchOrder           searchOrderMock;
    @Mock
    private SearchCondition       searchConditionMock;

    private final List<SearchOrder> searchOrders = new ArrayList<>();

    @Override
    protected SearchOrderRepository getRepo() {
        return repoMock;
    }

    @Override
    protected SearchOrder getEntity() {
        return searchOrderMock;
    }

    @Override
    public void specificSetUp() {
        service = new JooqSearchOrderService(repoMock, userRepoMock);

        searchOrders.add(searchOrderMock);
        searchOrders.add(searchOrderMock);
    }

    @Override
    public void specificTearDown() {
        verifyNoMoreInteractions(repoMock, filterMock, paginationContextMock, searchOrderMock, searchConditionMock);
    }

    @Test
    public void findingById_withFoundEntity_returnsOptionalOfIt() {
        Long id = 7L;
        when(repoMock.findById(id)).thenReturn(searchOrderMock);

        Optional<SearchOrder> optSearchOrder = service.findById(id);
        assertThat(optSearchOrder.isPresent()).isTrue();
        assertThat(optSearchOrder.get()).isEqualTo(searchOrderMock);

        verify(repoMock).findById(id);
        verifyAudit(1);
    }

    @Test
    public void findingById_withNotFoundEntity_returnsOptionalEmpty() {
        Long id = 7L;
        when(repoMock.findById(id)).thenReturn(null);

        assertThat(service
            .findById(id)
            .isPresent()).isFalse();

        verify(repoMock).findById(id);
    }

    @Test
    public void findingByFilter_delegatesToRepo() {
        when(repoMock.findPageByFilter(filterMock, paginationContextMock)).thenReturn(searchOrders);
        assertThat(service.findPageByFilter(filterMock, paginationContextMock)).isEqualTo(searchOrders);
        verify(repoMock).findPageByFilter(filterMock, paginationContextMock);
        verifyAudit(2);
    }

    @Test
    public void countingByFilter_delegatesToRepo() {
        when(repoMock.countByFilter(filterMock)).thenReturn(3);
        assertThat(service.countByFilter(filterMock)).isEqualTo(3);
        verify(repoMock).countByFilter(filterMock);
    }

    @Test
    public void savingOrUpdating_withSearchOrderWithNullId_hasRepoAddTheSearchOrder() {
        when(searchOrderMock.getId()).thenReturn(null);
        when(repoMock.add(searchOrderMock)).thenReturn(searchOrderMock);
        assertThat(service.saveOrUpdate(searchOrderMock)).isEqualTo(searchOrderMock);
        verify(repoMock).add(searchOrderMock);
        verify(searchOrderMock).getId();
        verifyAudit(1);
    }

    @Test
    public void savingOrUpdating_withSearchOrderWithNonNullId_hasRepoUpdateTheSearchOrder() {
        when(searchOrderMock.getId()).thenReturn(17L);
        when(repoMock.update(searchOrderMock)).thenReturn(searchOrderMock);
        assertThat(service.saveOrUpdate(searchOrderMock)).isEqualTo(searchOrderMock);
        verify(repoMock).update(searchOrderMock);
        verify(searchOrderMock).getId();
        verifyAudit(1);
    }

    @Test
    public void savingOrUpdatingSearchCondition_withConditionWithNullId_delegatesAddingToRepo() {
        final long searchOrderId = 3;
        when(searchConditionMock.getSearchConditionId()).thenReturn(null);
        when(repoMock.addSearchCondition(searchConditionMock, searchOrderId, LC)).thenReturn(searchConditionMock);
        assertThat(service.saveOrUpdateSearchCondition(searchConditionMock, searchOrderId, LC)).isEqualTo(
            searchConditionMock);
        verify(repoMock).addSearchCondition(searchConditionMock, searchOrderId, LC);
        verify(searchConditionMock).getSearchConditionId();
    }

    @Test
    public void savingOrUpdatingSearchCondition_withConditionWithId_delegatesUpdatingToRepo() {
        final long searchOrderId = 3;
        when(searchConditionMock.getSearchConditionId()).thenReturn(17L);
        when(repoMock.updateSearchCondition(searchConditionMock, searchOrderId, LC)).thenReturn(searchConditionMock);
        assertThat(service.saveOrUpdateSearchCondition(searchConditionMock, searchOrderId, LC)).isEqualTo(
            searchConditionMock);
        verify(repoMock).updateSearchCondition(searchConditionMock, searchOrderId, LC);
        verify(searchConditionMock).getSearchConditionId();
    }

    @Test
    public void deleting_withNullEntity_doesNothing() {
        service.remove(null);
        verify(repoMock, never()).delete(anyLong(), anyInt());
    }

    @Test
    public void deleting_withEntityWithNullId_doesNothing() {
        when(searchOrderMock.getId()).thenReturn(null);

        service.remove(searchOrderMock);

        verify(searchOrderMock).getId();
        verify(repoMock, never()).delete(anyLong(), anyInt());
    }

    @Test
    public void deleting_withEntityWithNormalId_delegatesToRepo() {
        when(searchOrderMock.getId()).thenReturn(3L);
        when(searchOrderMock.getVersion()).thenReturn(33);

        service.remove(searchOrderMock);

        verify(searchOrderMock, times(2)).getId();
        verify(searchOrderMock, times(1)).getVersion();
        verify(repoMock, times(1)).delete(3L, 33);
    }

    @Test
    public void removingSearchConditionWithId_withNullId_doesNothing() {
        service.removeSearchConditionWithId(null);
        verify(repoMock, never()).deleteSearchConditionWithId(anyLong());
    }

    @Test
    public void removingSearchConditionWithId_delegatesToRepo() {
        long id = 3L;
        service.removeSearchConditionWithId(id);
        verify(repoMock, times(1)).deleteSearchConditionWithId(id);
    }

}

File: core/core-persistence-jooq/src/test/kotlin/ch/difty/scipamato/core/persistence/search/JooqSearchOrderServiceTest.kt
package ch.difty.scipamato.core.persistence.search

import ch.difty.scipamato.common.persistence.paging.PaginationContext
import ch.difty.scipamato.core.entity.search.SearchCondition
import ch.difty.scipamato.core.entity.search.SearchOrder
import ch.difty.scipamato.core.entity.search.SearchOrderFilter
import ch.difty.scipamato.core.persistence.AbstractServiceTest
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeFalse
import org.amshove.kluent.shouldBeTrue
import org.junit.jupiter.api.Test

@Suppress("UsePropertyAccessSyntax")
internal class JooqSearchOrderServiceTest : AbstractServiceTest<Long, SearchOrder, SearchOrderRepository>() {

    override val repo = mockk<SearchOrderRepository>(relaxed = true) {
        every { delete(any(), any()) } returns entity
    }
    private val filterMock = mockk<SearchOrderFilter>()
    private val paginationContextMock = mockk<PaginationContext>()
    override val entity = mockk<SearchOrder>(relaxed = true)
    private val searchConditionMock = mockk<SearchCondition>()

    private var service = JooqSearchOrderService(repo, userRepoMock)

    private val searchOrders = listOf(entity, entity)

    public override fun specificTearDown() {
        confirmVerified(repo, filterMock, paginationContextMock, entity, searchConditionMock)
    }

    @Test
    fun findingById_withFoundEntity_returnsOptionalOfIt() {
        val id = 7L
        every { repo.findById(id) } returns entity
        auditFixture()

        val optSearchOrder = service.findById(id)
        optSearchOrder.isPresent.shouldBeTrue()
        optSearchOrder.get() shouldBeEqualTo entity

        verify { repo.findById(id) }
        verify { entity == entity }
        verify { entity.toString() }
        verifyAudit(1)
    }

    @Test
    fun findingById_withNotFoundEntity_returnsOptionalEmpty() {
        val id = 7L
        every { repo.findById(id) } returns null
        service.findById(id).isPresent.shouldBeFalse()
        verify { repo.findById(id) }
    }

    @Test
    fun findingByFilter_delegatesToRepo() {
        every { repo.findPageByFilter(filterMock, paginationContextMock) } returns searchOrders
        auditFixture()
        service.findPageByFilter(filterMock, paginationContextMock) shouldBeEqualTo searchOrders
        verify { repo.findPageByFilter(filterMock, paginationContextMock) }
        verify { entity.toString() }
        verifyAudit(2)
    }

    @Test
    fun countingByFilter_delegatesToRepo() {
        every { repo.countByFilter(filterMock) } returns 3
        service.countByFilter(filterMock) shouldBeEqualTo 3
        verify { repo.countByFilter(filterMock) }
    }

    @Test
    fun savingOrUpdating_withSearchOrderWithNullId_hasRepoAddTheSearchOrder() {
        every { entity.id } returns null
        every { repo.add(entity) } returns entity
        auditFixture()
        service.saveOrUpdate(entity) shouldBeEqualTo entity
        verify { repo.add(entity) }
        verify { entity.id }
        verify { entity == entity }
        verify { entity.toString() }
        verifyAudit(1)
    }

    @Test
    fun savingOrUpdating_withSearchOrderWithNonNullId_hasRepoUpdateTheSearchOrder() {
        every { entity.id } returns 17L
        every { repo.update(entity) } returns entity
        auditFixture()
        service.saveOrUpdate(entity) shouldBeEqualTo entity
        verify { repo.update(entity) }
        verify { entity.id }
        verify { entity == entity }
        verify { entity.toString() }
        verifyAudit(1)
    }

    @Test
    fun savingOrUpdatingSearchCondition_withConditionWithNullId_delegatesAddingToRepo() {
        val searchOrderId: Long = 3
        every { searchConditionMock.searchConditionId } returns null
        every { repo.addSearchCondition(searchConditionMock, searchOrderId, LC) } returns searchConditionMock
        service.saveOrUpdateSearchCondition(searchConditionMock, searchOrderId, LC) shouldBeEqualTo searchConditionMock
        verify { repo.addSearchCondition(searchConditionMock, searchOrderId, LC) }
        verify { searchConditionMock.searchConditionId }
        verify { searchConditionMock == searchConditionMock }
        verify { searchConditionMock.toString() }
    }

    @Test
    fun savingOrUpdatingSearchCondition_withConditionWithId_delegatesUpdatingToRepo() {
        val searchOrderId: Long = 3
        every { searchConditionMock.searchConditionId } returns 17L
        every { repo.updateSearchCondition(searchConditionMock, searchOrderId, LC) } returns searchConditionMock
        service.saveOrUpdateSearchCondition(searchConditionMock, searchOrderId, LC) shouldBeEqualTo searchConditionMock
        verify { repo.updateSearchCondition(searchConditionMock, searchOrderId, LC) }
        verify { searchConditionMock.searchConditionId }
        verify { searchConditionMock == searchConditionMock }
        verify { searchConditionMock.toString() }
    }

    @Test
    fun deleting_withNullEntity_doesNothing() {
        service.remove(null)
        verify(exactly = 0) { repo.delete(any(), any()) }
    }

    @Test
    fun deleting_withEntityWithNullId_doesNothing() {
        every { entity.id } returns null
        service.remove(entity)
        verify { entity.id }
        verify(exactly = 0) { repo.delete(any(), any()) }
    }

    @Test
    fun deleting_withEntityWithNormalId_delegatesToRepo() {
        every { entity.id } returns 3L
        every { entity.version } returns 33

        service.remove(entity)

        verify(exactly = 2) { entity.id }
        verify(exactly = 1) { entity.version }
        verify(exactly = 1) { repo.delete(3L, 33) }
    }

    @Test
    fun removingSearchConditionWithId_withNullId_doesNothing() {
        service.removeSearchConditionWithId(null)
        verify(exactly = 0) { repo.deleteSearchConditionWithId(any()) }
    }

    @Test
    fun removingSearchConditionWithId_delegatesToRepo() {
        val id = 3L
        service.removeSearchConditionWithId(id)
        verify(exactly = 1) { repo.deleteSearchConditionWithId(id) }
    }

    companion object {
        private const val LC = "de"
    }
}
